enumerable.rb

```diff
@@ -0,0 +1,179 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    arr = self if instance_of?(Array)
+    arr = to_a if instance_of?(Range)
+    arr = flatten if instance_of?(Hash)
+
+    count = 0
+    while count < arr.length
+      yield(arr[count])
+      count += 1
+    end
+    arr
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    arr = self if instance_of?(Array)
+    arr = to_a if instance_of?(Range)
+    arr = flatten if instance_of?(Hash)
+
+    count = 0
+    while count < arr.length
+      yield(arr[count], count)
+      count += 1
+    end
+    arr
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    new_arr = []
+    my_each { |it| new_arr << it if yield(it) }
+    new_arr
+  end
+
+  def my_all?(param = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == false }
+      return true
+    elsif param.nil?
+      my_each { |num| return false if num.nil? || num == false }
+    elsif !param.nil? && (param.is_a? Class)
+      my_each { |num| return false if num.class != param }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      my_each { |_num| return false unless arg.match(n) }
+    else
+      my_each { |_num| return false if n != param }
+    end
+    true
+  end
+
+  def my_any?(param = nil)
+    if block_given?
+      my_each { |item| return true if yield(item) }
+      return true
+    elsif param.nil?
+      my_each { |num| return true if num.nil? || num == true }
+    elsif !param.nil? && (param.is_a? Class)
+      my_each { |num| return true if num.instance_of?(param) }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      my_each { |_num| return true unless arg.match(num) }
+    else
+      my_each { |_num| return true if num != param }
+    end
+    true
+  end
+
+  def my_none?(param = nil)
+    if !block_given? && param.nil?
+      my_each { |num| return true if num }
+      return false
+    end
+
+    if !block_given? && !param.nil
+
+      if param.is_a?(Class)
+        my_each { |num| return false if num.instance_of?(param) }
+        return true
+      end
+
+      if param.instance_of?(Regexp)
+        my_each { |num| return false if arg.match(num) }
+        return true
+      end
+
+      my_each { |num| return false if num == param }
+      true
+    end
+    my_any? { |it| return false if yield(it) }
+    true
+  end
+
+  def my_count(num = nil)
+    arr = instance_of?(Array) ? self : to_a
+    return arr.length unless block_given? || num
+
+    return arr.my_select { |item| item == num }.length if num
+
+    arr.my_select { |item| yield(item) }.length
+  end
+
+  def my_map(par_ = nil)
```

Comment:
enumerables #my_map returns an Enumerator if no block is given
```
     Failure/Error: expect(array.my_map).to be_an(Enumerator)
```

enumerables #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```
      Failure/Error: expect(array.my_map(my_proc) { |num| num < 10 }).to eq array.map(&my_proc)
      
        expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
             got: [4, 1, 1, 6, 4, 4, 7, 7, 0, 0, 7, 3, 7, 5, 5, 5, 4, 1, 2, 0, 4, 0, 8, 7, 5, 0, 3, 3, 2, 6, 5, 2, 3, 8..., 0, 6, 3, 6, 5, 0, 8, 1, 8, 8, 3, 6, 3, 1, 7, 1, 3, 1, 3, 6, 6, 4, 7, 1, 1, 3, 0, 1, 4, 2, 3, 6, 8]
```

# 
### `my_map` should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)).